### PR TITLE
Add new 'combineBy' transform type

### PR DESCRIPTION
**Pull request by ghost**
Hello! I have been obsessed with hydra over the past week or so and I've been having a lot of fun using it, but there are a few things I really wish I could do, so I thought I might try to extend hydra for this purpose.

There are a lot of functions where I would like to be able to use a generator as the `amount` parameter (like `blend`), so this extends the glsl generator to accommodate for this use-case. 

Started with just `blendBy` as proof of concept to discuss.

Example:

```js
s1.initCam()

src(s1).thresh(0.5, 0.2).out(o1)

osc(100,0.1,0.2).blendBy(voronoi(), o1).out()
```
---

~~I'm okay with javascript, but unsure how to test these changes?~~ (Figured it out!)

### CODE REVIEW

**Comment by ghost** [comment]
It works!

```js
s1.initVideo("https://media.giphy.com/media/26ufplp8yheSKUE00/giphy.mp4", {})
  
src(s1).thresh(0.9, 0.3).invert().out(o1)

osc(100,0.1,0.2).blendBy(noise(), o1).out()
```

image

**Comment by ghost** [comment]
A possible extension/follow-up would be to allow generators in place of any parameter.
Breaking change (if redundant functions get removed), but would make hydra a lot more powerful.

File: src/generator-factory.js [hunk 1]
```diff
@@ -135,11 +149,11 @@ const typeLookup = {
 function processGlsl(obj) {
   let t = typeLookup[obj.type]
   if(t) {
-  let baseArgs = t.args.map((arg) => arg).join(", ")
+  let baseArgs = t.args.map((arg) => `${arg.type} ${arg.name}`).join(", ")
   // @todo: make sure this works for all input types, add validation
   let customArgs = obj.inputs.map((input) => `${input.type} ${input.name}`).join(', ')
   let args = `${baseArgs}${customArgs.length > 0 ? ', '+ customArgs: ''}`
-//  console.log('args are ', args)
+  console.log('args are ', args)
```
Comment (ghost) [inline src/generator-factory.js]: Re-comment this out

File: src/generator-factory.js [hunk 2]
```diff
@@ -148,14 +162,14 @@ function processGlsl(obj) {
...
 
-  // add extra input to beginning for backward combatibility @todo update compiler so this is no longer necessary
-    if(obj.type === 'combine' || obj.type === 'combineCoord') obj.inputs.unshift({
-        name: 'color',
-        type: 'vec4'
-      })
+    t.args.slice(1).forEach((input) => {
+      obj.inputs.unshift(input);
+      console.log(input)
+    });
```
Comment (ghost) [inline src/generator-factory.js]: Remove this `console.log`, also consider something like:

```js
obj.inputs = t.args.concat(obj.inputs);
```

**Comment by geikha** [comment]
Hi! It's lovely to see new ideas for the parsing! On the combineBy,  you can already do this with Hydra. I see that we should document this specific use case. The solution is using a `vec4` argument!

```js
setFunction({
  name: 'blendBy',
  type: 'combine',
  inputs: [
    {
      type: 'vec4',
      name: 'amount',
      default: 0.5,
    }
  ],
  glsl:
`   return _c0*(vec4(1.0)-amount)+_c1*amount;`
})

osc().blendBy(voronoi(),noise(1).thresh(0)).out()
```

On the other hand, the usage of making everythign modulatable is something we've talked in the past and could be a great feature to add!

**Comment by ghost** [comment]
> On the other hand, the usage of making everything modulatable is something we've talked in the past and could be a great feature to add!

Right on! It didn't work when I tried it using the normal blend function, but from the way the generator worked it seemed like it should have. The type mismatch definitely explains it!

I'll close this PR and see about making everything modulate-able.

Thank you for your time and kind words :)

**Pull request closed without merging**